client/lib/widgets/sharing_form.dart

```diff
@@ -0,0 +1,171 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+import 'package:studybuddy/services/database.dart' as database;
+import 'package:studybuddy/widgets/user_tile.dart';
+import 'package:studybuddy/routes/routes.dart' as routes;
+
+class SharingForm extends StatefulWidget {
+  /// {@macro add_todo_popup_card}
+  const SharingForm({
+    Key? key,
+    required this.course,
+    required this.isOwner,
+  }) : super(key: key);
+  final QueryDocumentSnapshot<Object?> course;
+  final bool isOwner;
+
+  @override
+  State<SharingForm> createState() => _SharingFormState();
+}
+
+var currentUser = FirebaseAuth.instance.currentUser;
+
+class _SharingFormState extends State<SharingForm> {
+  final TextEditingController _namecontroller = TextEditingController();
+  late Future<dynamic> courseUserPermissions;
+  late bool isLoading;
+
+  void _refreshPermList() {
+    setState(() {
+      courseUserPermissions = database.getCoursePermList(widget.course.id);
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    isLoading = false;
+    courseUserPermissions = database.getCoursePermList(widget.course.id);

```

Comment:
since the course object contains the permissions on initial load and we only care about changing when it's refreshed, don't think we need this first call on the init? we could instead initialize `courseUserPermissions` here with something like `Future<dynamic>.value(widget.course.roles)` to avoid the extra database call

Reply:
courses are not set up as streams yet so there are inconsistency issues. will fix this in a different pr as it doesn't pertain directly to course sharing